Complete Edge Detect Module

HarshKhandeparkar opened the pull request:
Fixes #176

- Hysteresis is implemented(it slows down the browsers very significantly)
- Black background
- General refactoring

### It Works!!

Old:
![edge-detect 6](https://user-images.githubusercontent.com/34770591/52835887-6e385700-310e-11e9-8002-eed37e7211db.png)

- This output had a lot of transparency(alpha = 0)
- Thick edges
- No hysteresis

New:
![edge-detect 5](https://user-images.githubusercontent.com/34770591/52835912-814b2700-310e-11e9-9486-ca203531a4e4.png)

- Black background
- No transparency
- Pure white edges
- Thin edges

##### Note: The default high and low threshold ratio values are not necessarily meant to be the best. They are to be found empirically for different cases.

##### Note: Higher high threshold takes less time to process but it's lossier (some edges are neglected).

##### Note: Lower low/high threshold takes very long to process and can render a browser unresponsive.

##### Note: Low threshold ratio can be greater than the high threshold ratio since it is multiplied to the high threshold value. i.e `low_threshold_ratio = low_threshold_value/high_threshold_value` and `high_threshold_ratio = high_threshold_value/maximum_gradient`

##### Note: Here value is the actual value which is compared to the gradients and ratio is a relative ratio of different values.

### It is slow though 
![too-slow-edge-detect](https://user-images.githubusercontent.com/34770591/52835976-ca9b7680-310e-11e9-8bc3-b820fbe6d688.gif)


It takes very long to iterate through all the edges. This algorithm/module should be mainly used on servers. This can benefit a lot from the REST API project which is under development. 

@publiclab/is-reviewers @jywarren 

Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

* [x] tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR -- or run tests locally with `npm test`
* [x] code is in uniquely-named feature branch and has no merge conflicts
* [x] PR is descriptively titled
* [x] ask `@publiclab/reviewers` and `@publiclab/is-reviewers` for help, in a comment below

****

## Thanks!

HarshKhandeparkar commented:
I have fixed the angle(i guess?) but I haven't built the dist as i did that in the browser. Can someone build the dist or atleast check out the changes locally?
@jywarren @tech4GT @publiclab/is-reviewers

jywarren commented:
Wow, lots of important work here. Can you link to a definition for hysteresis? And can you show a screenshot of how this worked before and how it's different now? Thanks @HarshKhandeparkar !!! Great work here!

HarshKhandeparkar commented:
@jywarren I don't really know the definition of hysterisis. What it essentially does is it removes weakEdgePixels if they are not connected to strong pixels otherwise preserves them.
There is a lot of info about it here

https://en.m.wikipedia.org/wiki/Canny_edge_detector

HarshKhandeparkar commented:
Btw this implementation of hysterisis doesn't check for diagonally connected strong and weak pixels because that really, really!! Slows down the browser. Thanks!

HarshKhandeparkar commented:
@jywarren I won't be able to provide a screenshot of how this works now. Atleast not today. I haven't even built the dist files as I updated some files in the browser. Can you try it out locally? I will upload a screenshot of how it worked earlier.

![screenshot_20190214-222138](https://user-images.githubusercontent.com/34770591/52802938-ef5c0380-30a6-11e9-9c4c-ff40a716c8ea.jpg)

This is the old output. Now the edges should be white and background black. Also the edges (should?) be thinner. Even I haven't ckecked out the new output to be honest (as I changed files in the browser)

HarshKhandeparkar commented:
@jywarren @publiclab/is-reviewers everything seems to work. I have linked a few pics and GIFs above, please check them out. Please let me know if this is the desired output and also review the code.
@tech4GT

tech4GT commented:
This looks great Harsh, I'll see if I can optimize anything here, but so far so good :)

tech4GT commented on package.json:
```diff
@@ -39,7 +39,6 @@
     "jquery": "^3.3.1",
     "jsdom": "^13.1.0",
     "jsqr": "^1.1.1",
-    "lodash": "^4.17.5",
```
@HarshKhandeparkar Any reason for removing lodash?

HarshKhandeparkar replied:
No I will add it again. I did not realize that I removed it. 😊

HarshKhandeparkar commented:
@tech4GT if you are trying this out locally, then will you please be able to reinstall lodash and rebase for me?

tech4GT commented:
Oh I'll try it out tonight! I'll do it then if you want.

HarshKhandeparkar commented:
Thanks! I will be busy for a few days 😊

On Sat 16 Feb, 2019, 4:08 PM Varun Gupta <notifications@github.com wrote:

> Oh I'll try it out tonight! I'll do it then if you want.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/publiclab/image-sequencer/pull/771#issuecomment-464334862>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AhKOnyXK_ISLePjNOF2345NYD25R3QcHks5vN9-cgaJpZM4a7gVP>
> .
>

HarshKhandeparkar commented:
@tech4GT have you tested this out yet?

HarshKhandeparkar commented:
@jywarren can you please merge my PRs before the other ones since I will not be able to touch my PC until 16th of april. I don't know if I will get time past that date as well.

jywarren commented:
OK, this is really impressive. A few questions:

1. how slow is it? Can you run it with the new benchmarking tests and compare it to the existing edge detection? Is it much slower than that?
2. can you add your notes which are really useful from the PR description to the module documentation here: https://github.com/publiclab/image-sequencer/blob/main/docs/MODULES.md#edge-detect-module
3. if hysteresis is what's making it slow, is it possible to toggle it on and off using an option? 
4. can you try running it on the image in this post? https://publiclab.org/questions/jlev/02-01-2018/how-can-i-identify-bits-of-plastic-from-the-beach-in-an-image

No rush on this. I'd love if someone else could help out here too!

HarshKhandeparkar commented:
I think it will timeout in most cases. I don't think it should be
benchmarked each time. Maybe it should be made optional?

On Mon 4 Mar, 2019, 9:15 PM Jeffrey Warren, <notifications@github.com>
wrote:

> OK, this is really impressive. A few questions:
>
>    1. how slow is it? Can you run it with the new benchmarking tests and
>    compare it to the existing edge detection? Is it much slower than that?
>    2. can you add your notes which are really useful from the PR
>    description to the module documentation here:
>    https://github.com/publiclab/image-sequencer/blob/main/docs/MODULES.md#edge-detect-module
>    3. if hysteresis is what's making it slow, is it possible to toggle it
>    on and off using an option?
>    4. can you try running it on the image in this post?
>    https://publiclab.org/questions/jlev/02-01-2018/how-can-i-identify-bits-of-plastic-from-the-beach-in-an-image
>
> No rush on this. I'd love if someone else could help out here too!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/publiclab/image-sequencer/pull/771#issuecomment-469300970>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AhKOnzwiVDYcLApw4UeB-5tcSnBIcRxfks5vTT-0gaJpZM4a7gVP>
> .
>

HarshKhandeparkar commented:
The module is slow in general. Not because of hysterisis but yes. Hysteresis makes it even slower. Earlier it didn't timeout. Now it does. Should it be made an option?

jywarren commented:
Should hysteresis be made an option? Maybe, esp. if it is significantly lengthening the runtime. 

Are the change to make background black and only black or white pixels (and no greyscale) which I like a lot here, are they part of hysteresis, or if we let people turn off hysteresis, they'll still get black and white images on a black background?

HarshKhandeparkar commented:
No, hysterisis will only remove any superfine edges created out of noise. In most cases hysteresis will not make a lot of difference. The black backround will still be there.

HarshKhandeparkar commented:
I have a question. Who had created the module originally?

jywarren commented:
@tech4GT did!

jywarren commented:
OK so would it be difficult to make hysteresis toggle-able?

HarshKhandeparkar commented:
Nope. It is just a func which can be called when required.

jywarren commented:
Super, let's do that then! Thanks, @HarshKhandeparkar !

jywarren commented on src/modules/EdgeDetect/EdgeUtils.js:
```diff
@@ -1,174 +1,201 @@
+  }
+  nonMaxSupress(pixels, grads, angles);
+  doubleThreshold(pixels, highThresholdRatio, lowThresholdRatio, grads, strongEdgePixels, weakEdgePixels);
+  hysteresis(strongEdgePixels, weakEdgePixels);
```
Here, we'd say `if (options.hysteresis)` ?

And in the info.json we'd set that as an option defaulting to true i guess?

HarshKhandeparkar replied:
Yes. Exactly. It can default to false as well. Maybe we can default it to false since this module can crash old browsers if used by mistake. The browser will simply freeze. Maybe we can default it to true after alternate thread processing is implemented. Also the option it gets from the info.json can be a string and not a boolean and can also have a capital letter at the start so it might get itself messed up. Using
```js
options.hysteresis.toLowerCase() == 'true'
```
Is beeter I guess.

jywarren commented:
that sounds great, defaulting to false!

On Mon, Mar 4, 2019 at 1:09 PM Harsh Khandeparkar <notifications@github.com>
wrote:

> *@HarshKhandeparkar* commented on this pull request.
> ------------------------------
>
> In src/modules/EdgeDetect/EdgeUtils.js
> <https://github.com/publiclab/image-sequencer/pull/771#discussion_r262178395>
> :
>
> >      }
> -    let magY = 0.0;
> -    for (let a = 0; a < 3; a++) {
> -        for (let b = 0; b < 3; b++) {
> +  }
> +  nonMaxSupress(pixels, grads, angles);
> +  doubleThreshold(pixels, highThresholdRatio, lowThresholdRatio, grads, strongEdgePixels, weakEdgePixels);
> +  hysteresis(strongEdgePixels, weakEdgePixels);
>
> Yes. Exactly. It can default to false as well. Maybe we can default it to
> false since this module can crash old browsers if used by mistake. The
> browser will simply freeze. Maybe we can default it to true after alternate
> thread processing is implemented. Also the option it gets from the
> info.json can be a string and not a boolean and can also have a capital
> letter at the start so it might get itself messed up. Using
>
> options.hysteresis.toLowerCase() == 'true'
>
> Is beeter I guess.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/publiclab/image-sequencer/pull/771#discussion_r262178395>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABfJ2LHS80S6iDuxAPWjqzC1FiPG-alks5vTWE8gaJpZM4a7gVP>
> .
>

HarshKhandeparkar commented:
@tech4GT is this PR complete? Is my work proper? Does it actually complete the module or is something remaining? Is hysteresis correctly implemented? Anything?

cc @jywarren

jywarren commented:
Ah, did you manage to add the addition so that hysteresis defaults to false, and is an option (with checkbox?) Thank you!!!

HarshKhandeparkar commented:
I will add everything as soon as I get time so I was asking for all the required suggestions.

HarshKhandeparkar commented:
@jywarren I have rebased and made all the changes. A demo will be available [here](https://harshkhandeparkar.github.io/image-sequencer/examples/new)

HarshKhandeparkar commented:
Please try to merge this before the others. Thank you!

HarshKhandeparkar commented:
Oh yea, please be sure to clear the cache.

HarshKhandeparkar commented:
An updated link will be A demo will be available [here](https://harshkhandeparkar.github.io/image-sequencer/examples/legacy). Sorry for the inconvenience.

HarshKhandeparkar commented:
@jywarren I think this is ready to merge. I have checked everything. It works. Below is hysteresis v/s non-hysteresis comparison(left image is without hysteresis).

![Screenshot from 2019-03-17 11-12-40](https://user-images.githubusercontent.com/34770591/54485819-cac19a00-48a5-11e9-90b4-ebeb25f850ca.png)


### Thank You!

cc @tech4GT @publiclab/is-reviewers

Harshithpabbati commented:
Hey Harsh the UI looks different in mobile.

HarshKhandeparkar commented:
Hello harshith, try clearing the cache and do not use the new UI. And css will be fixed by @aashna27

Harshithpabbati commented:
![phone](https://user-images.githubusercontent.com/43822585/54485886-91d5f500-48a6-11e9-9dc5-1323bfc19cf4.jpg)
check this out. I cleared the cache.

HarshKhandeparkar commented:
I removed your image because it was covering the screen. That issue will be fixed by @aashna27 in #867

Harshithpabbati commented:
If you want I can make the changes in css.

HarshKhandeparkar commented:
Atleast not now. As I turned off my pc and won't be able to do anything. You can open a pr in my gh-pages branch if you want

HarshKhandeparkar commented:
This current PR focuses on edgeDetect so css isn't much important. If you still want to improve the demo, you can. Thank you.

jywarren commented on src/modules/EdgeDetect/Module.js:
```diff
@@ -7,6 +7,7 @@ module.exports = function edgeDetect(options, UI) {
   options.blur = options.blur || defaults.blur;
   options.highThresholdRatio = options.highThresholdRatio || defaults.highThresholdRatio;
   options.lowThresholdRatio = options.lowThresholdRatio || defaults.lowThresholdRatio;
+  options.hystereis = options.hysteresis || defaults.hysteresis;
```
Oops, i think there's a missing `s`? `hysteresis`?

HarshKhandeparkar replied:
I have made a typo where it is implemented as well so it doesn't make any difference. Two negatives make a +ve lol😂. I make a lot of typos. We can open an fto for this later. A good to welcome new contributors.

jywarren commented:
And noting that it passed in almost the same amount of time:

this branch:     Module edge-detect ran in: 2672 milliseconds
`main`:    Module edge-detect ran in: 2709 milliseconds

HarshKhandeparkar commented:
It takes a lot less time without hysteresis. Maybe it ran faster here as there is no different gradient and there are less pixels to be preserved. I.e binary image. I'm not sure though.

HarshKhandeparkar commented:
For the tests, maybe we can open an issue for creation of looks-same tests for all modules.

jywarren commented:
Hi Harsh, i'm OK with merging this without a test, but I think in general we should be transitioning to requiring a test with every module before merging it, rather than doing it in follow-up, as a matter of policy. What do you think?

HarshKhandeparkar commented:
Yes sure. Thanks. I can add tests to all the remaining modules once I am free and maybe add the new policy to the contributing file.

jywarren commented:
Great, thank you! Really nice work here!

HarshKhandeparkar commented:
Thanks a lot!!! @Harshithpabbati this was just merged. Try running tge update script when you get time. Let's see if the changes are visible there.

Harshithpabbati commented:
Sure @harshkhandeparkar I will run it now.
